src/main/scala/firrtl/passes/CheckHighForm.scala

```diff
@@ -270,7 +281,11 @@ trait CheckHighFormLike { this: Pass =>
         case NoInfo => minfo
         case x      => x
       }
-      s.foreach(checkName(info, mname, names))
+      val canBeReference = s match {
+        case _: Stop | _: Print | _: Verification => false
```

Comment:
Maybe we can have a sealed trait in IR.scala that codifies this property.

Reply:
This used to be `HasName` since every `Statement` that had a name would also be able to be referenced.
Should I add a `CanBeReference` trait that extends from `HasName`?

Reply:
Maybe a self-type constraint, to keep the class hierarchy shallow? @jackkoenig has better-informed opinions about this sort of thing

Reply:
What if we make the side-effecting statements just extend `HasName` and not `IsDeclaration`? Alternatively, if we want to keep the meaning that `IsDeclaration` means "Annotatable", I think David's idea with a shallower type hierarchy is good so that it'll be more obvious:
```scala
case class DefWire(info: Info, name: String, tpe: Type) extends Statement with IsDeclaration with CanBeReference { ...
```
Verbose, yes, but no tracing through type hierarchies.

Reply:
I can add the `CanBeReference` trait. Side-effecting statements should extend `IsDeclaration` since that makes them part of the module level namespace.